src/rhsm/utils.py

```diff
@@ -297,6 +298,102 @@ def wrapper(*args, **kwargs):
     return wrapper
 
 
+def singleton(cls: type) -> type:
+    """Decorate a class to make it singleton.
+
+    This decorator is inherited: subclasses will be singletons as well,
+    but they won't be the same singleton as their parent.
+    """
+
+    new = cls.__new__
```

Comment:
I would use name `__orig_new` or something like this to emphasize that it is original __new__ method. It would be good to use type hint in this case.